data_model/derive/src/lib.rs

```diff
@@ -341,3 +342,82 @@ pub fn partially_tagged_serialize_derive(input: TokenStream) -> TokenStream {
 pub fn partially_tagged_deserialize_derive(input: TokenStream) -> TokenStream {
     partially_tagged::impl_partially_tagged_deserialize(&parse_macro_input!(input))
 }
+
+/// Derive macro for `HasOrigin`.
+///
+/// Works only with single unnamed field enums.
+///
+/// # Attributes
+///
+/// ## Container attributes
+///
+/// ### `#[has_origin(origin = Type)]`
+///
+/// Required attribute. Used to determine type of `Origin` in `HasOrigin` trait.
+///
+/// ## Field attributes
+///
+/// ### `#[has_origin(ident => expr)]`
+///
+/// This attribute is used to determine how to extract origin id from enum variant.
+/// By default variant is assumed to by origin id.
+///
+/// # Examples
+///
+/// ```
+/// use iroha_data_model_derive::{IdOrdEqHash, HasOrigin};
+/// use iroha_data_model::prelude::{Identifiable, HasOrigin};
+///
+///
+/// #[derive(HasOrigin, Clone, Debug)]
+/// #[has_origin(origin = Layer)]
+/// pub enum LayerEvent {
+///     #[has_origin(sub_layer_event => &sub_layer_event.origin_id().parent_id)]
+///     SubLayer(SubLayerEvent),
+///     Created(LayerId),
+/// }
+///
+/// #[derive(HasOrigin, Clone, Debug)]
+/// #[has_origin(origin = SubLayer)]
+/// pub enum SubLayerEvent {
+///     Created(SubLayerId),
+/// }
+///
+/// #[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq, Hash)]
+/// pub struct LayerId {
+///     name: u32,
+/// }
+///
+/// #[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq, Hash)]
+/// pub struct SubLayerId {
+///     name: u32,
+///     parent_id: LayerId,
+/// }
+///
+/// #[derive(IdOrdEqHash, Debug, Clone)]
+/// #[id(type = "LayerId")]
```

Comment:
After the `IdOrdEqHash` refactor done by @mversic, these inner attributes are no longer needed, as it is sufficient for a field just to be named `id`. I was surprised the doc test still passed, and upon checking, it seems the macro defaults to looking at the field name currently and silently ignores the unneeded attr. That might be not what we want.

Reply:
Yeah, on container level macros are just ignored.
And this is not only place where we have such kind of behavior (ignoring unused attributes).

Reply:
I feel like that would make more sense, yeah. Because such attributes are attached when we want *certain* behavior, and silently ignoring it should become known to the end user. So be it a change in API or some mistaken usage, it's better to be explicit. If you create an issue, that would be much appreciated.